plugins/X3DReader/X3DReader.py

```diff
@@ -0,0 +1,902 @@
+from UM.Scene.GroupDecorator import GroupDecorator
+from UM.Job import Job
+from math import pi, sin, cos, sqrt
+import numpy
+
+
+EPSILON = 0.000001 # So very crude. :(
+
+try:
+    import xml.etree.cElementTree as ET
+except ImportError:
+    import xml.etree.ElementTree as ET
+    
+# TODO: preserve the structure of scenes that contain several objects
+# Use CADPart, for example, to distinguish between separate objects    
+    
+DEFAULT_SUBDIV = 16 # Default subdivision factor for spheres, cones, and cylinders
+
+class Shape:
+    def __init__(self, v, f, ib, n):
+        self.verts = v
+        self.faces = f
+        # Those are here for debugging purposes only
+        self.index_base = ib 
+        self.name = n
+        
+class X3DReader(MeshReader):
+    def __init__(self):
+        super().__init__()
+        self._supported_extensions = [".x3d"]
+        self._namespaces = {}
+    
+    # Main entry point
+    # Reads the file, returns a SceneNode (possibly with nested ones), or None
+    def read(self, file_name):
+        try:
+            self.defs = {}
+            self.shapes = []
+            
+            tree = ET.parse(file_name)
+            xml_root = tree.getroot()
+            
+            if xml_root.tag != "X3D":
+                return None
+
+            scale = 1000 # Default X3D unit it one meter, while Cura's is one millimeters            
+            if xml_root[0].tag == "head":
+                for head_node in xml_root[0]:
+                    if head_node.tag == "unit" and head_node.attrib.get("category") == "length":
+                        scale *= float(head_node.attrib["conversionFactor"])
+                        break 
+                xml_scene = xml_root[1]
+            else:
+                xml_scene = xml_root[0]
+                
+            if xml_scene.tag != "Scene":
+                return None 
+            
+            self.transform = Matrix()
+            self.transform.setByScaleFactor(scale)
+            self.index_base = 0
+            
+            # Traverse the scene tree, populate the shapes list
+            self.processChildNodes(xml_scene)
+            
+            if self.shapes:
+                bui = MeshBuilder()
+                bui.setVertices(numpy.concatenate([shape.verts for shape in self.shapes]))
+                bui.setIndices(numpy.concatenate([shape.faces for shape in self.shapes]))
+                bui.calculateNormals()
+                bui.setFileName(file_name)
+                    
+                scene = SceneNode()
+                scene.setMeshData(bui.build().getTransformed(Matrix()))
+                scene.setSelectable(True)
+                scene.setName(file_name)
+            else:
+                return None
+            
+        except Exception as e:
+            Logger.log("e", "exception occured in x3d reader: %s", e)
+
+        try:
+            boundingBox = scene.getBoundingBox()
+            boundingBox.isValid()
```

Comment:
This line doesn't do anything, unless bounding box is None (which causes the exception to be raised).
The exception is also far to broad. If you want an exeption that is that broad, document why and let the logger write the exception somewhere.